// eslint-watch/build/eslint/help.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = eslintHelp;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var _cli = require('./cli');

var _cli2 = _interopRequireDefault(_cli);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var logger = (0, _logger2.default)('eslint-help');
logger.debug('Loaded');

var namedOption = /^--/;

function parseNo(option, str) {
  if (!str) return;

  var cmd = str.replace('--', '');
  if (/no-/.test(cmd)) {
    logger.debug('Parsing no option', str);
    cmd = cmd.replace('no-', '');
    option.default = 'true';
  }
  option.option = cmd;
  return option;
}

function parseDouble(arr) {
  var description = _lodash2.default.without(arr.slice(2), '').join(' ');
  return {
    option: arr[0].replace('--', ''),
    type: 'Boolean',
    alias: arr[1].replace('--', ''),
    description: description
  };
}

function parseRegular(arr) {
  logger.debug('Parsing %s', arr[0]);
  if (!arr[0]) {
    return;
  }
  var optionText = arr[0];
  var type = arr[1] || 'Boolean';
  var option = {};
  option = parseNo(option, optionText);

  option.type = type;

  var helpText = _lodash2.default.without(arr, optionText, type, '');

  var description = helpText.join(' ');
  if (description) {
    option.description = description;
  }
  return option;
}

function parseAlias(arr) {
  var alias = arr[0];
  logger.debug('Alias found: %s', alias);
  var option = parseRegular(_lodash2.default.without(arr, alias));

  if (alias) {
    option.alias = alias.replace('-', '');
  }
  return option;
}

function createOption(arr) {
  var option = void 0;

  if (namedOption.test(arr[0]) && namedOption.test(arr[1])) {
    // no alias defaulted boolean
    option = parseDouble(arr);
  } else if (namedOption.test(arr[0]) && !namedOption.test(arr[1])) {
    // just a no alias
    option = parseRegular(arr);
  } else {
    // aliased or other
    option = parseAlias(arr);
  }
  var isEmpty = _lodash2.default.isEmpty(option);
  return isEmpty ? undefined : option;
}

function parseHelp(helpText) {
  var helpArr = helpText.split('\n');
  var newArr = [];
  var previousLine = [];
  _lodash2.default.each(helpArr, function (row, index) {
    if (index <= 2) {
      return;
    } else {
      var str = row.replace(',', '');
      var arr = str.trim().split(' ');
      if (str.indexOf('-') >= 0 && previousLine[0] !== '') {
        var option = createOption(arr);
        if (option && option.option !== 'format' && option.option !== 'help') {
          newArr.push(option);
        }
      }
      previousLine = arr;
    }
  });
  return newArr;
}

function eslintHelp() {
  logger.debug('Executing help');
  var result = (0, _cli2.default)(['--help'], { stdio: [process.stdin, null, process.stderr] });
  if (!result.message) {
    throw new Error('Help text not received from Eslint.');
  }
  var eslintOptions = parseHelp(result.message);
  return eslintOptions;
};
module.exports = exports['default'];
